panic: cannot inline references within correlated subqueries

I was investigating #28816, trying to pare down the query and ran into a panic in `InlineProjections`.

query:
```sql
query I
SELECT
  subq_0.c0 AS c0
FROM (SELECT 1 AS c0, 2 as c1) AS subq_0
WHERE
  1
  >= CASE
    WHEN subq_0.c1 IS NOT NULL
    THEN pg_catalog.extract(
      CAST(
        CASE
        WHEN
        (
            EXISTS(
              SELECT
                ref_1.config_yaml AS c0,
                ref_1.config_yaml AS c1,
                subq_0.c0 AS c2,
                ref_1.config_yaml AS c3
              FROM
                crdb_internal.zones AS ref_1
              WHERE
                subq_0.c0 IS NOT NULL
              LIMIT
                52
            )
          )
        THEN pg_catalog.version()
        ELSE pg_catalog.version()
        END
          AS TEXT
      ),
      CAST(pg_catalog.current_date() AS DATE)
    )
    ELSE 1
    END
LIMIT
  107
----
```

```
panic: cannot inline references within correlated subqueries [recovered]
	panic: cannot inline references within correlated subqueries [recovered]
	panic: panic while executing 1 statements: SELECT _._ AS _ FROM (SELECT _ AS _, _ AS _) AS _ WHERE _ >= CASE WHEN _._ IS NOT _ THEN extract(CAST(CASE WHEN (EXISTS (SELECT _._ AS _, _._ AS _, _._ AS _, _._ AS _ FROM _._ AS _ WHERE _._ IS NOT _ LIMIT _)) THEN version() ELSE version() END AS TEXT), CAST(current_date() AS DATE)) ELSE _ END LIMIT _; caused by cannot inline references within correlated subqueries

goroutine 405 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc4201c5900, 0x2777b20, 0xc420dc0780, 0x1f399c0, 0x274a230)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:675 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc4201c5900, 0x2777b20, 0xc420dc0780)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:414 +0x61
panic(0x1f399c0, 0x274a230)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build.func1(0xc420c6a998)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:130 +0x99
panic(0x1f399c0, 0x274a230)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0xc400000010, 0xc420c68d60)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:120 +0x5cd
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1738, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:181 +0x9f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff003200000015, 0xc4201c5dd0, 0xc420c69f50, 0xc42044d598, 0xc420c68e78)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0x1a00000015, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1878, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:181 +0x9f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff006500000017, 0xc4201c5dd0, 0xc420c69f50, 0xc420d6e000, 0xffff002b00000012)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0x17, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1918, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:191 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff006400000018, 0xc4201c5dd0, 0xc420c69f50, 0xc420d6ed30, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0xc400000018, 0xc420d6e000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e19b8, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:181 +0x9f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff006300000019, 0xc4201c5dd0, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0x19, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1af8, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:191 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff00670000001b, 0xc4201c5dd0, 0xc420c69f50, 0xc420d6e000, 0xffff004a00000020)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0x1b, 0x20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1b98, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:181 +0x9f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff00650000001c, 0xc4201c5dd0, 0xc420c69f50, 0xc420d6e000, 0xffff002b00000012)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0x1c, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1c38, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:191 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff00640000001d, 0xc4201c5dd0, 0xc420c69f50, 0x3, 0x1d0000003a)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0x1d, 0x1d0000003a)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1cd8, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:181 +0x9f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff003a0000001e, 0xc4201c5dd0, 0xc420c69f50, 0xffff003200000015, 0xffff003200650017)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0x1e, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Expr).Replace(0xc4201e1d78, 0xc420d6e000, 0xc420c69f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.go:191 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExprView.Replace(0xc420d6e000, 0xffff00330000001f, 0xc4201c5dd0, 0xc420c69f50, 0xc420c69e78, 0x16db6fa)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr_view.go:201 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections.func1(0xc40000001f, 0x176d097)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:165 +0x49f
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).InlineProjections(0xc4201c6888, 0x50000001f, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline.go:168 +0x155
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructSelect(0xc4201c6878, 0x1f00000006, 0x1f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:850 +0x1c55
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildFrom(0xc420336540, 0xc42086e120, 0xc4202ad840, 0xc421150400, 0xc4211505d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:527 +0x592
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelectClause(0xc420336540, 0xc420c6c480, 0x0, 0x0, 0x0, 0xc421150400, 0x30)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:429 +0x56
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelect(0xc420336540, 0xc420442dc0, 0xc421150400, 0x21fd580)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:387 +0x1a6
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildStmt(0xc420336540, 0x277aea0, 0xc420442dc0, 0xc421150400, 0x4)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:177 +0xf1
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build(0xc420336540, 0xc400000000, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:135 +0x128
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc4201c5d38, 0x2777be0, 0xc42086e510, 0x277aea0, 0xc420442dc0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:378 +0x287
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optionallyUseOptimizer(0xc4201c5d38, 0x2777be0, 0xc42086e510, 0x0, 0x0, 0xc4207de011, 0x4, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/planner.go:534 +0xdb
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc4201c5900, 0x2777be0, 0xc42086e510, 0x277aea0, 0xc420442dc0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:751 +0x192
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc4201c5900, 0x2777be0, 0xc42086e510, 0x277aea0, 0xc420442dc0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:401 +0xb82
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc4201c5900, 0x2777be0, 0xc42086e510, 0x277aea0, 0xc420442dc0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc4201c5900, 0x2777b20, 0xc420dc0780, 0xc4202c0e10, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1079 +0x2186
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc4201c0000, 0x2777b20, 0xc420dc0780, 0xc4207de011, 0x4, 0xc4207de030, 0x4, 0x0, 0x0, 0x275dea0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:416 +0x1bb
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc4201c0000, 0x2777b20, 0xc420dc0780, 0xc4205ea700, 0x5400, 0x15000, 0xc420612180, 0xc4202c0e10, 0xc4202c0e00, 0xc420471180, ...)
```

--- a/pkg/sql/opt/memo/logical_props_builder.go
+++ b/pkg/sql/opt/memo/logical_props_builder.go
@@ -1059,8 +1059,9 @@ func (b *logicalPropsBuilder) buildScalarProps(ev ExprView) props.Logical {
 		return logical
 	}
 
-	// By default, derive OuterCols and CanHaveSideEffects from all children, both
-	// relational and scalar.
+	// By default, derive OuterCols and CanHaveSideEffects from all children,
+	// both relational and scalar. Derive HasCorrelatedSubquery from scalar
+	// children.
 	for i, n := 0, ev.ChildCount(); i < n; i++ {
 		childLogical := &ev.childGroup(i).logical
 
@@ -1068,6 +1069,10 @@ func (b *logicalPropsBuilder) buildScalarProps(ev ExprView) props.Logical {
 
 		if childLogical.CanHaveSideEffects() {
 			scalar.CanHaveSideEffects = true
+		}
+
+		if childLogical.Scalar != nil && childLogical.Scalar.HasCorrelatedSubquery {
+			scalar.HasCorrelatedSubquery = true
 		}
 	}
 
@@ -1114,12 +1119,17 @@ func (b *logicalPropsBuilder) buildScalarProps(ev ExprView) props.Logical {
 		funcOpDef := ev.Private().(*FuncOpDef)
 		if funcOpDef.Properties.Impure {
 			// Impure functions can return different value on each call.
-			logical.Scalar.CanHaveSideEffects = true
+			scalar.CanHaveSideEffects = true
 		}
 
 	case opt.DivOp:
 		// Division by zero error is possible.
-		logical.Scalar.CanHaveSideEffects = true
+		scalar.CanHaveSideEffects = true
+
+	case opt.SubqueryOp, opt.ExistsOp, opt.AnyOp:
+		if !scalar.OuterCols.Empty() {
+			scalar.HasCorrelatedSubquery = true
+		}
 	}
 
 	return logical

--- a/pkg/sql/opt/norm/custom_funcs.go
+++ b/pkg/sql/opt/norm/custom_funcs.go
@@ -346,6 +346,12 @@ func (c *CustomFuncs) CanHaveZeroRows(group memo.GroupID) bool {
 	return c.mem.GroupProperties(group).Relational.Cardinality.CanBeZero()
 }
 
+// HasCorrelatedSubquery returns true if the given scalar group contains a
+// subquery within its subtree that has at least one outer column.
+func (c *CustomFuncs) HasCorrelatedSubquery(group memo.GroupID) bool {
+	return c.LookupScalar(group).HasCorrelatedSubquery
+}
+
 // ----------------------------------------------------------------------
 //
 // Private extraction functions

--- a/pkg/sql/opt/norm/decorrelate.go
+++ b/pkg/sql/opt/norm/decorrelate.go
@@ -28,8 +28,14 @@ import (
 // subquery needs to be hoisted up into its parent query as part of query
 // decorrelation.
 func (c *CustomFuncs) HasHoistableSubquery(group memo.GroupID) bool {
-	// Lazily calculate and store the HasHoistableSubquery value.
+	// Don't bother traversing the expression tree if there is no correlated
+	// subquery.
 	scalar := c.LookupScalar(group)
+	if !scalar.HasCorrelatedSubquery {
+		return false
+	}
+
+	// Lazily calculate and store the HasHoistableSubquery value.
 	if scalar.IsAvailable(props.HasHoistableSubquery) {
 		return scalar.Rule.HasHoistableSubquery
 	}

--- a/pkg/sql/opt/norm/inline.go
+++ b/pkg/sql/opt/norm/inline.go
@@ -25,13 +25,20 @@ import (
 //
 //   SELECT x+1, x+2, y FROM a
 //
-// hasDuplicateRefs would be true for the Projections expression, since the x
+// HasDuplicateRefs would be true for the Projections expression, since the x
 // column is referenced twice.
 //
 // Correlated subqueries are disallowed since it introduces additional
 // complexity for a case that's not important for inlining (correlated
 // subqueries are hoisted to a higher context anyway).
 func (c *CustomFuncs) HasDuplicateRefs(target memo.GroupID) bool {
+	// Don't bother traversing the expression tree if there is a correlated
+	// subquery.
+	scalar := c.LookupScalar(target)
+	if scalar.HasCorrelatedSubquery {
+		return true
+	}
+
 	var refs opt.ColSet
 
 	// When a column reference is found, add it to the refs set. If the set
@@ -41,9 +48,10 @@ func (c *CustomFuncs) HasDuplicateRefs(target memo.GroupID) bool {
 	findDupRefs = func(group memo.GroupID) bool {
 		expr := c.f.mem.NormExpr(group)
 		if !expr.IsScalar() {
-			// Don't try to count references within correlated subqueries.
+			// We know that this is not a correlated subquery since
+			// scalar.HasCorrelatedSubquery was already checked above.
 			// Uncorrelated subqueries never have references.
-			return !c.OuterCols(group).Empty()
+			return false
 		}
 
 		switch expr.Operator() {
@@ -116,7 +124,7 @@ func (c *CustomFuncs) InlineProjections(target, projections memo.GroupID) memo.G
 		expr := c.f.mem.NormExpr(child)
 		if !expr.IsScalar() {
 			if !c.OuterCols(child).Empty() {
-				// Should have prevented this in hasDuplicateRefs/canInline.
+				// Should have prevented this in HasDuplicateRefs/HasCorrelatedSubquery.
 				panic("cannot inline references within correlated subqueries")
 			}
 			return child

--- a/pkg/sql/opt/norm/rules/inline.opt
+++ b/pkg/sql/opt/norm/rules/inline.opt
@@ -35,7 +35,7 @@
         $input:*
         $projections:* & (CanInline $projections)
     )
-    $filter:*
+    $filter:* & ^(HasCorrelatedSubquery $filter)
 )
 =>
 (Project

--- a/pkg/sql/opt/norm/testdata/rules/inline
+++ b/pkg/sql/opt/norm/testdata/rules/inline
@@ -118,6 +118,118 @@ semi-join
  └── filters [type=bool, outer=(2,6)]
       └── (x - i) > (i * i) [type=bool, outer=(2,6)]
 
+exec-ddl
+CREATE TABLE crdb_internal.zones (
+    zone_id INT NOT NULL,
+    cli_specifier STRING NULL,
+    config_yaml BYTES NOT NULL,
+    config_protobuf BYTES NOT NULL
+)
+----
+TABLE zones
+ ├── zone_id int not null
+ ├── cli_specifier string
+ ├── config_yaml bytes not null
+ ├── config_protobuf bytes not null
+ ├── rowid int not null (hidden)
+ └── INDEX primary
+      └── rowid int not null (hidden)
+
+# Regression test for #28827. Ensure that inlining is not applied when there
+# is a correlated subquery in the filter.
+norm
+SELECT
+  subq_0.c0 AS c0
+FROM (SELECT 1 AS c0, 2 as c1) AS subq_0
+WHERE
+  1
+  >= CASE
+    WHEN subq_0.c1 IS NOT NULL
+    THEN pg_catalog.extract(
+      CAST(
+        CASE
+        WHEN
+        (
+            EXISTS(
+              SELECT
+                ref_1.config_yaml AS c0,
+                ref_1.config_yaml AS c1,
+                subq_0.c0 AS c2,
+                ref_1.config_yaml AS c3
+              FROM
+                crdb_internal.zones AS ref_1
+              WHERE
+                subq_0.c0 IS NOT NULL
+              LIMIT
+                52
+            )
+          )
+        THEN pg_catalog.version()
+        ELSE pg_catalog.version()
+        END
+          AS TEXT
+      ),
+      CAST(pg_catalog.current_date() AS DATE)
+    )
+    ELSE 1
+    END
+LIMIT
+  107
+----
+project
+ ├── columns: c0:1(int!null)
+ ├── cardinality: [0 - 1]
+ ├── side-effects
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── select
+      ├── columns: c0:1(int!null) c1:2(int!null)
+      ├── cardinality: [0 - 1]
+      ├── side-effects
+      ├── key: ()
+      ├── fd: ()-->(1,2)
+      ├── project
+      │    ├── columns: c0:1(int!null) c1:2(int!null)
+      │    ├── cardinality: [1 - 1]
+      │    ├── key: ()
+      │    ├── fd: ()-->(1,2)
+      │    ├── values
+      │    │    ├── cardinality: [1 - 1]
+      │    │    ├── key: ()
+      │    │    └── tuple [type=tuple]
+      │    └── projections
+      │         ├── const: 1 [type=int]
+      │         └── const: 2 [type=int]
+      └── filters [type=bool, outer=(1,2), side-effects]
+           └── le [type=bool, outer=(1,2), side-effects]
+                ├── case [type=int, outer=(1,2), side-effects]
+                │    ├── true [type=bool]
+                │    ├── when [type=int, outer=(1,2), side-effects]
+                │    │    ├── c1 IS NOT NULL [type=bool, outer=(2)]
+                │    │    └── function: extract [type=int, outer=(1), side-effects]
+                │    │         ├── cast: TEXT [type=string, outer=(1)]
+                │    │         │    └── case [type=string, outer=(1)]
+                │    │         │         ├── true [type=bool]
+                │    │         │         ├── when [type=string, outer=(1)]
+                │    │         │         │    ├── exists [type=bool, outer=(1)]
+                │    │         │         │    │    └── limit
+                │    │         │         │    │         ├── columns: config_yaml:5(bytes!null)
+                │    │         │         │    │         ├── outer: (1)
+                │    │         │         │    │         ├── cardinality: [0 - 52]
+                │    │         │         │    │         ├── select
+                │    │         │         │    │         │    ├── columns: config_yaml:5(bytes!null)
+                │    │         │         │    │         │    ├── outer: (1)
+                │    │         │         │    │         │    ├── scan zones
+                │    │         │         │    │         │    │    └── columns: config_yaml:5(bytes!null)
+                │    │         │         │    │         │    └── filters [type=bool, outer=(1), constraints=(/1: (/NULL - ]; tight)]
+                │    │         │         │    │         │         └── c0 IS NOT NULL [type=bool, outer=(1), constraints=(/1: (/NULL - ]; tight)]
+                │    │         │         │    │         └── const: 52 [type=int]
+                │    │         │         │    └── function: version [type=string]
+                │    │         │         └── function: version [type=string]
+                │    │         └── function: current_date [type=date, side-effects]
+                │    └── const: 1 [type=int]
+                └── const: 1 [type=int]
+
 # --------------------------------------------------
 # InlineProjectInProject
 # --------------------------------------------------

--- a/pkg/sql/opt/props/logical.go
+++ b/pkg/sql/opt/props/logical.go
@@ -262,6 +262,14 @@ type Scalar struct {
 	// comment for Logical.CanHaveSideEffects.
 	CanHaveSideEffects bool
 
+	// HasCorrelatedSubquery is true if the scalar expression tree contains a
+	// subquery having one or more outer columns. The subquery can be a Subquery,
+	// Exists, or Any operator. These operators need to be hoisted out of scalar
+	// expression trees and turned into top-level apply joins. This property makes
+	// detection fast and easy so that the hoister doesn't waste time searching
+	// subtrees that don't contain subqueries.
+	HasCorrelatedSubquery bool
+
 	// Constraints is the set of constraints deduced from a boolean expression.
 	// For the expression to be true, all constraints in the set must be
 	// satisfied.

